[Serve] [Doc] Update intro page (#27735)
ray-project/ray: doc/source/serve/index.md

Serve is built on top of Ray, so it easily scales to many machines and offers flexible scheduling support such as fractional GPUs so you can share resources and serve many machine learning models at low cost.

## Quickstart

Install Ray Serve and its dependencies:

```bash
pip install "ray[serve]"
```

> To run this example, install the following: ``pip install ray["serve"]``

Reviewer: Looks redundant with the above? Also the quotes should be on the outside, `"ray[serve]"`
<img width="605" alt="Screen Shot 2022-08-10 at 1 12 20 PM" src="https://user-images.githubusercontent.com/5459654/184010987-567cebc4-807c-4cac-ad48-9411f15fc42d.png">

```{literalinclude} doc_code/quickstart.py
:language: python
```

:::{tabbed} More examples
For more examples, select from the tabs.
:::

:::{tabbed} Model composition

> In this example we demonstrate Serve's model composition API allowing you to express complex computation graph and deploy them as a unit.

Reviewer: ```suggestion
In this example we demonstrate Serve's model composition API allowing you to express a complex computation graph and deploy it as a unit.
```

> In this example we demonstrate Serve's model composition API allowing you to express a complex computation graph and deploy it as a unit.

Reviewer 2: ```suggestion
In this example we demonstrate Serve's model composition API allowing you to express a complex computation graph and deploy it as a Serve application.
```

```
:::

> :::{tabbed} HuggingFace model

Reviewer: ```suggestion
:::{tabbed} Hugging Face model
```

:animate: fade-in-slide-down

Ray Serve brings the scalability and parallelism of these hosted offerings to
your own infrastructure. You can use our [cluster launcher](cluster-cloud)
to deploy Ray Serve to all major public clouds, K8s, as well as on bare-metal, on-premise machines.

Compared to these offerings, Ray Serve lacks a unified user interface and functionality to
let you manage the lifecycle of the models, visualize it's performance, etc. Ray
Serve primarily focuses on model serving and provides the primitives for you to
build your own ML platform on top.
:::

:::{dropdown} Seldon, KServe, Cortex
:animate: fade-in-slide-down

You can develop Ray Serve on your laptop, deploy it on a dev box, and scale it out
to multiple machines or a K8s cluster with minimal or no changes to code. It's a lot
easier to get started with when you don't need to provision and manage a K8s cluster.
When it's time to deploy, you can use our [Kubernetes Operator](https://ray-project.github.io/kuberay/guidance/rayservice/)
to transparently put your Ray Serve application in K8s.
:::

:::{dropdown} BentoML, Comet.ml, MLflow
:animate: fade-in-slide-down

> Ray Serve is a special purpose distributed model server built for large scale applications.
> This means we can work with any model packaging and registry format.
> Many of the tools are focused on serving one ML framework-specific type of a model. By contrast, Ray Serve is framework-agnostic and allows you to build an end-to-end machine learning application with its unique model composition API and advanced autoscaling capabilities.
> :::
> 
> We truly believe Serve is unique as it gives you end to end control
> over the API while delivering scalability and high performance. To achieve
> Serve's feature offerings, you often need to glue together multiple
> frameworks like Tensorflow Serving, SageMaker, or even roll your own
> batching server.

Reviewer 3: ```suggestion
Many of these tools are focused on serving models for one specific ML framework. In contrast, Ray Serve is framework-agnostic and focuses on large-scale model serving in general.
As such, Ray Serve works with any model packaging and registry format.
Ray Serve also provides key features for building end-to-end machine learning applications, including model composition and autoscaling.
:::

We truly believe Serve is unique as it gives you end-to-end control
over the API while delivering scalability and high performance. To achieve
Serve's feature offerings with other tools, you would need to glue together multiple
frameworks like Tensorflow Serving and SageMaker, or even roll your own
batching server.
```

What tools are "the tools" referring to? I'm going to assume it's the ones that are listed at the top. Reordered to make this clearer.

Also don't know what "the API" is referring to, or why it's significant to have end-to-end control over it. Serve's API?

What's a "batching server"?

:::

:::{dropdown} ML engineer
:animate: fade-in-slide-down

Serve helps you scale out your deployment and runs them reliably and efficiently to save costs. With Serve's first-class model composition API, you can combine models together with business logic and build end-to-end user-facing applications. Additionally, Serve runs natively on Kubernetes with minimal operation overhead.
:::

:::{dropdown} ML platform engineer
:animate: fade-in-slide-down

Serve specializes in scalable and reliable ML model serving. As such, it can be an important plug-and-play component of your ML platform stack.
Serve supports arbitrary Python code and therefore integrates well with the MLOps ecosystem. You can use it with model optimizers (ONNX, TVM), model monitoring systems (Seldon Alibi, Arize), model registries (MLFlow, Weights and Biases), machine learning frameworks (XGBoost, Scikit-learn), data app UIs (Gradio, Streamlit), and Web API frameworks (FastAPI, gRPC).

:::

## How does Serve compare to ...

:::{dropdown} TFServing, TorchServe, ONNXRuntime
:animate: fade-in-slide-down

Ray Serve is *framework agnostic*, so you can use it alongside any other Python framework or library.

> We believe data scientists should not be bounded to a particular machine learning framework.

Reviewer 3: ```suggestion
We believe data scientists should not be bound to a particular machine learning framework.
```

They should be empowered to use the best tool available for the job.

Compared to these framework-specific solutions, Ray Serve doesn't perform any model-specific optimizations to make your ML model run faster. However, you can still optimize the models yourself
and run them in Ray Serve. For example, you can run a model compiled by
[PyTorch JIT](https://pytorch.org/docs/stable/jit.html) or [ONNXRuntime](https://onnxruntime.ai/).
:::

:::{dropdown} AWS SageMaker, Azure ML, Google Vertex AI
:animate: fade-in-slide-down

As an open-source project, Ray Serve brings the scalability and reliability of these hosted offerings to your own infrastructure.
You can use the Ray [cluster launcher](cluster-cloud) to deploy Ray Serve to all major public clouds, K8s, as well as on bare-metal, on-premise machines.

Ray Serve is not a full-fledged ML Platform.
Compared to these other offerings, Ray Serve lacks the functionality for
managing the lifecycle of your models, visualizing their performance, etc. Ray
Serve primarily focuses on model serving and providing the primitives for you to
build your own ML platform on top.

> If you are looking for end-to-end ML pipeline framework from data processing to serving, check out [Ray AI Runtime](air).

Reviewer 3: ```suggestion
If you are looking for end-to-end ML pipeline framework that can handle everything from data processing to serving, check out [Ray AI Runtime](air).
```

:::

:::{dropdown} Seldon, KServe, Cortex
:animate: fade-in-slide-down

You can develop Ray Serve on your laptop, deploy it on a dev box, and scale it out
to multiple machines or a Kubernetes cluster, all with minimal or no changes to code. It's a lot
easier to get started with when you don't need to provision and manage a K8s cluster.
When it's time to deploy, you can use our [Kubernetes Operator](kuberay-quickstart)
to transparently deploy your Ray Serve application to K8s.
:::

:::{dropdown} BentoML, Comet.ml, MLflow
:animate: fade-in-slide-down

Many of these tools are focused on serving and scaling models independently.
In contrast, Ray Serve is framework-agnostic and focuses on model composition.
As such, Ray Serve works with any model packaging and registry format.
Ray Serve also provides key features for building production-ready machine learning applications, including best-in-class autoscaling and naturally integrating with business logic.
:::

We truly believe Serve is unique as it gives you end-to-end control

> over the API while delivering scalability and high performance. To achieve

Reviewer 3: Still not sure what "the API" means :)

:::

## How can Serve help me as a...

:::{dropdown} Data scientist
:animate: fade-in-slide-down

> Serve is easy to use. You can test your models (and your entire deployment graph) on your local machine before deploying it to production on a cluster. You don't need to know heavyweight Kubernetes concepts or cloud providers to use Serve.

Reviewer 2: ```suggestion
Serve is easy to use. You can test your models (and your entire deployment graph) on your local machine before deploying it to production on a cluster. You don't need to know heavyweight Kubernetes concepts or cloud configurations to use Serve.
```

:::

:::{dropdown} ML engineer
:animate: fade-in-slide-down

> Serve helps you scale out the deployment and runs them reliably, efficiently, and cheaply to save costs. With Serve's first-class model composition API, you can combine models together and build end-to-end application. Additionally, Serve natively runs on Kubernetes with minimal operation overhead.

Reviewer 2: ```suggestion
Serve helps you scale out the deployment and runs them reliably and efficiently to save costs. With Serve's first-class model composition API, you can combine models together and build end-to-end applications. Additionally, Serve natively runs on Kubernetes with minimal operation overhead.
```
efficiently and cheaply is redundant (and cheaply isn't a word :))

:::

:::{dropdown} ML Platform engineer
:animate: fade-in-slide-down

Serve specializes in ML model serving. As such, it can be an important and reliable component as part of your ML Platform stack.
Serve supports arbitrary Python code and therefore integrates well with the MLOps ecosystem. You can use it with model optimizers (ONNX, TVM), model monitoring systems (Seldon Alibi, Arize), model registries (MLFlow, Weights and Biases), machine learning frameworks (XGBoost, Scikit-learn), data app UIs (Gradio, Streamlit), and Web API frameworks (FastAPI, gRPC).

:::

> ## How does Serve compare to X

Reviewer: ```suggestion
## How does Serve compare to...
```
Nit, to make it consistent with the previous header "How can Serve help me as a…"

:::{dropdown} TFServing, TorchServe, ONNXRuntime
:animate: fade-in-slide-down

Ray Serve is *framework agnostic*, so you can use any Python framework and libraries.
We believe data scientists are not bounded a particular machine learning framework.
They use the best tool available for the job.

Compared to these framework-specific solutions, Ray Serve doesn't perform any optimizations to make your ML model run faster. However, you can still optimize the models yourself

> and run them in Ray Serve: for example, you can run a model compiled by

Reviewer 2: ```suggestion
and run them in Ray Serve. For example, you can run a model compiled by
```

> [PyTorch JIT](https://pytorch.org/docs/stable/jit.html).

Reviewer 2: also mention onnx maybe?

:::

:::{dropdown} AWS SageMaker, Azure ML, Google Vertex AI
:animate: fade-in-slide-down

Ray Serve brings the scalability and parallelism of these hosted offerings to
your own infrastructure. You can use our [cluster launcher](cluster-cloud)
to deploy Ray Serve to all major public clouds, K8s, as well as on bare-metal, on-premise machines.

Compared to these offerings, Ray Serve lacks a unified user interface and functionality to
let you manage the lifecycle of the models, visualize it's performance, etc. Ray
Serve primarily focuses on model serving and provides the primitives for you to
build your own ML platform on top.
:::

:::{dropdown} Seldon, KServe, Cortex
:animate: fade-in-slide-down

You can develop Ray Serve on your laptop, deploy it on a dev box, and scale it out
to multiple machines or a K8s cluster with minimal or no changes to code. It's a lot
easier to get started with when you don't need to provision and manage a K8s cluster.

> When it's time to deploy, you can use our [Kubernetes Operator](https://ray-project.github.io/kuberay/guidance/rayservice/)

Reviewer 2: we should link to my prod section instead of kuberay after it's merged

:::

## How can Serve help me as a...

:::{dropdown} Data scientist
:animate: fade-in-slide-down

Serve is easy to use. You can test your models (and your entire deployment graph) on your local machine before deploying it to production on a cluster. You don't need to know heavyweight Kubernetes concepts or cloud providers to use Serve.

:::

:::{dropdown} ML engineer
:animate: fade-in-slide-down

> Serve helps you scale out the deployment and run them reliably in an efficient, cost-saving manner. Serve offers a first-class model composition API to help you combine models together to build end-to-end application. Additionally, Serve natively run on Kubernetes which minimal operation overhead.

Reviewer 4: Serve helps you scale out the deployment and runs them reliably, efficiently, and cheaply to save costs. With Serve's first-class model composition API, you can combine models together and build end-to-end application. Additionally, Serve natively runs on Kubernetes with minimal operation overhead.

:::

:::{dropdown} ML Platform engineer
:animate: fade-in-slide-down

Serve specializes in ML model serving and can be a reliably component in your ML Platform stack.

> Serve supports arbitrary Python code and therefore integrates well with the MLOps ecosystem. You can use it with model optimizers (ONNX, TVM), model monitoring systems (Seldon Alibi, Arize), model registries (MLFlow, Weights and Biases), machine learning frameworks (XGBoost, Scikit-learn), data app UIs (Gradio, Streamlit), and Web API frameworks (FastAPI, gRPC).

Reviewer 4: Serve specializes in ML model serving. As such, it can be an important and reliable component as part of your ML Platform stack.

Reviewer 2: don't capitalize "Platform"

:::

## How does Serve compare to X

:::{dropdown} TFServing, TorchServe, ONNXRuntime
:animate: fade-in-slide-down

Ray Serve is *framework agnostic*, you can use any Python framework and libraries.

> We believe data scientists are not bounded a particular machine learning framework.

Reviewer 4: Ray Serve is *framework agnostic*, so you can use any Python framework and libraries.

They use the best tool available for the job.

Compared to these framework specific solution, Ray Serve doesn't perform any optimizations

> to make your ML model run faster. However, you can still optimize the models yourself

Reviewer 4: Compared to these framework-specific solutions, Ray Serve doesn't perform any optimizations
to make your ML model run faster.

and run them in Ray Serve: for example, you can run a model compiled by
[PyTorch JIT](https://pytorch.org/docs/stable/jit.html).
:::

:::{dropdown} AWS SageMaker, Azure ML, Google Vertex AI
:animate: fade-in-slide-down

Ray Serve brings the scalability and parallelism of these hosted offering to

> your own infrastructure. You can use our [cluster launcher](cluster-cloud)

Reviewer 4: Ray Serve brings the scalability and parallelism of these hosted offerings to
your own infrastructure. You can use Ray's [cluster launcher](cluster-cloud)

to deploy Ray Serve to all major public clouds, K8s, as well as on bare-metal, on-premise machines.

Compared to these offerings, Ray Serve lacks a unified user interface and functionality
let you manage the lifecycle of the models, visualize it's performance, etc. Ray

> Serve focuses on just model serving and provides the primitives for you to

Reviewer 4: Compared to these offerings, Ray Serve lacks a unified user interface and functionality to
let you manage the lifecycle of the models, visualize its performance, etc. Ray
Serve primarily focuses on model serving and ...

build your own ML platform on top.
:::

:::{dropdown} Seldon, KServe, Cortex
:animate: fade-in-slide-down

You can develop Ray Serve on your laptop, deploy it on a dev box, and scale it out

> to multiple machines or K8s cluster without changing one lines of code. It's a lot

Reviewer 4: "without changing one lines of code" 

I would say "with minimal or no changes to code."

> easier to get started with when you don't need to provision and manage K8s cluster.

Reviewer: ```suggestion
easier to get started with when you don't need to provision and manage a K8s cluster.
```

> When it's time to deploy, you can use the our native Kubernetes Operator

Reviewer: ```suggestion
When it's time to deploy, you can use our native Kubernetes Operator
```

Reviewer: Ideally we would add a link here

to transparently put your Ray Serve application in K8s.
:::

:::{dropdown} BentoML, Comet.ml, MLflow
:animate: fade-in-slide-down

Ray Serve is a special purpose distributed model server built for large scale applications.
This means we can work with any model packaging and registry format. Many of the tools are
focused on serving one models. Ray Serve is built for end-to-end machine learning application

> in mind with our unique model composition API and advanced autoscaling capabilities.

Reviewer 4: Many of the tools are focused on serving one ML framework-specific type of a model. By contrast, Ray Serve is framework-agnostic and allows you to build an end-to-end machine learning application with its unique model composition API and advanced autoscaling capabilities.

:::

We truly believe Serve is unique as it gives you end to end control
over the API while delivering scalability and high performance. To achieve
something like what Serve offers, you often need to glue together multiple
frameworks like Tensorflow Serving, SageMaker, or even roll your own

> batching server.

Reviewer 4: We truly believe Serve is unique as it gives you end to end control
over the API while delivering scalability and high performance. To achieve
Serve's feature offerings, you often need to glue together multiple
frameworks like Tensorflow Serving, SageMaker, or even roll your own
batching server.

Reviewer 4: Might also want to add this blog post and meetup talk as links at the end

-[ Multi-model composition with Ray Serve deployment graphs](https://www.anyscale.com/blog/multi-model-composition-with-ray-serve-deployment-graphs)

- [Productionizing ML at scale with Ray Serve](https://www.anyscale.com/events/2022/04/14/productionizing-ml-at-scale-with-ray-serve)
